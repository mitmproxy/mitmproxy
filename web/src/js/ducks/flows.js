import { fetchApi } from "../utils"
import reduceStore from "./utils/store"
import * as storeActions from "./utils/store"
import Filt from "../filt/filt"
import { RequestUtils } from "../flow/utils"

export const ADD            = 'FLOWS_ADD'
export const UPDATE         = 'FLOWS_UPDATE'
export const REMOVE         = 'FLOWS_REMOVE'
export const RECEIVE        = 'FLOWS_RECEIVE'
export const SELECT         = 'FLOWS_SELECT'
export const SET_FILTER     = 'FLOWS_SET_FILTER'
export const SET_SORT       = 'FLOWS_SET_SORT'
export const SET_HIGHLIGHT  = 'FLOWS_SET_HIGHLIGHT'
export const REQUEST_ACTION = 'FLOWS_REQUEST_ACTION'


const defaultState = {
    highlight: null,
    filter: null,
    sort: { column: null, desc: false },
    selected: [],
    ...reduceStore(undefined, {})
}

export default function reduce(state = defaultState, action) {
    switch (action.type) {

        case ADD:
        case UPDATE:
        case REMOVE:
        case RECEIVE:
            let storeAction = storeActions[action.cmd](
                action.data,
                makeFilter(state.filter),
                makeSort(state.sort)
            )

            let selected = state.selected
            if(action.type === REMOVE && state.selected.includes(action.data)) {
                if(state.selected.length > 1){
                    selected = selected.filter(x => x !== action.data)
                } else {
                    selected = []
                    if (action.data in state.viewIndex && state.view.length > 1) {
                        let currentIndex = state.viewIndex[action.data],
                            nextSelection
                        if(currentIndex === state.view.length -1){ // last row
                            nextSelection = state.view[currentIndex - 1]
                        } else {
                            nextSelection = state.view[currentIndex + 1]
                        }
                        selected.push(nextSelection.id)
                    }
                }
            }

            return {
                ...state,
                selected,
                ...reduceStore(state, storeAction)
            }

        case SET_FILTER:
            return {
                ...state,
                filter: action.filter,
                ...reduceStore(state, storeActions.setFilter(makeFilter(action.filter), makeSort(state.sort)))
            }

        case SET_HIGHLIGHT:
            return {
                ...state,
                highlight: action.highlight
            }

        case SET_SORT:
            return {
                ...state,
                sort: action.sort,
                ...reduceStore(state, storeActions.setSort(makeSort(action.sort)))
            }

        case SELECT:
            return {
                ...state,
                selected: action.flowIds
            }

        default:
            return state
    }
}


const sortKeyFuns = {

    TLSColumn: flow => flow.request.scheme,

    PathColumn: flow => RequestUtils.pretty_url(flow.request),

    MethodColumn: flow => flow.request.method,

    StatusColumn: flow => flow.response && flow.response.status_code,

    TimeColumn: flow => flow.response && flow.response.timestamp_end - flow.request.timestamp_start,

    SizeColumn: flow => {
        let total = flow.request.contentLength
        if (flow.response) {
            total += flow.response.contentLength || 0
        }
        return total
    },
}

export function makeFilter(filter) {
    if (!filter) {
        return
    }
    return Filt.parse(filter)
}

export function makeSort({ column, desc }) {
    const sortKeyFun = sortKeyFuns[column]
    if (!sortKeyFun) {
        return
    }
    return (a, b) => {
        const ka = sortKeyFun(a)
        const kb = sortKeyFun(b)
        if (ka > kb) {
            return desc ? -1 : 1
        }
        if (ka < kb) {
            return desc ? 1 : -1
        }
        return 0
    }
}

export function setFilter(filter) {
    return { type: SET_FILTER, filter }
}

export function setHighlight(highlight) {
    return { type: SET_HIGHLIGHT, highlight }
}

export function setSort(column, desc) {
    return { type: SET_SORT, sort: { column, desc } }
}

export function selectRelative(flows, shift) {
    let currentSelectionIndex = flows.viewIndex[flows.selected[0]]
    let minIndex = 0
    let maxIndex = flows.view.length - 1
    let newIndex
    if (currentSelectionIndex === undefined) {
        newIndex = (shift < 0) ? minIndex : maxIndex
    } else {
        newIndex = currentSelectionIndex + shift
        newIndex = window.Math.max(newIndex, minIndex)
        newIndex = window.Math.min(newIndex, maxIndex)
    }
    let flow = flows.view[newIndex]
    return select(flow ? flow.id : undefined)
}


export function resume(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/resume`, { method: 'POST' })
}

export function resumeAll() {
    return dispatch => fetchApi('/flows/resume', { method: 'POST' })
}

export function kill(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/kill`, { method: 'POST' })
}

export function killAll() {
    return dispatch => fetchApi('/flows/kill', { method: 'POST' })
}


export function remove(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}`, { method: 'DELETE' })
}

export function duplicate(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/duplicate`, { method: 'POST' })
}

export function replay(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/replay`, { method: 'POST' })
}

export function revert(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/revert`, { method: 'POST' })
}

export function update(flow, data) {
    return dispatch => fetchApi.put(`/flows/${flow.id}`, data)
}

export function uploadContent(flow, file, type) {
    const body = new FormData()
    file       = new window.Blob([file], { type: 'plain/text' })
    body.append('file', file)
    return dispatch => fetchApi(`/flows/${flow.id}/${type}/content.data`, { method: 'POST', body })
}


export function clear() {
    return dispatch => fetchApi('/clear', { method: 'POST' })
}

export function download() {
    window.location = '/flows/dump'
    return { type: REQUEST_ACTION }
}

export function upload(file) {
    const body = new FormData()
    body.append('file', file)
    return dispatch => fetchApi('/flows/dump', { method: 'POST', body })
}


export function select(id) {
    return {
        type: SELECT,
        flowIds: id ? [id] : []
    }
}
