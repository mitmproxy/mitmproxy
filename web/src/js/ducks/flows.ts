import {fetchApi} from "../utils"
import * as store from "./utils/store"
import Filt from "../filt/filt"
import {Flow} from "../flow";
import FlowColumns from "../components/FlowTable/FlowColumns";

export const ADD = 'FLOWS_ADD'
export const UPDATE = 'FLOWS_UPDATE'
export const REMOVE = 'FLOWS_REMOVE'
export const RECEIVE = 'FLOWS_RECEIVE'
export const SELECT = 'FLOWS_SELECT'
export const SET_FILTER = 'FLOWS_SET_FILTER'
export const SET_SORT = 'FLOWS_SET_SORT'
export const SET_HIGHLIGHT = 'FLOWS_SET_HIGHLIGHT'

interface FlowSortFn extends store.SortFn<Flow> {
}

interface FlowFilterFn extends store.FilterFn<Flow> {
}


export interface FlowsState extends store.State<Flow> {
    highlight?: string,
    filter?: string,
    sort: { column?: keyof typeof FlowColumns, desc: boolean },
    selected: string[],
}

export const defaultState: FlowsState = {
    highlight: undefined,
    filter: undefined,
    sort: {column: undefined, desc: false},
    selected: [],
    ...store.defaultState
}

export default function reducer(state: FlowsState = defaultState, action): FlowsState {
    switch (action.type) {

        case ADD:
        case UPDATE:
        case REMOVE:
        case RECEIVE:
            let storeAction = store[action.cmd](
                action.data,
                makeFilter(state.filter),
                makeSort(state.sort)
            )

            let selected = state.selected
            if (action.type === REMOVE && state.selected.includes(action.data)) {
                if (state.selected.length > 1) {
                    selected = selected.filter(x => x !== action.data)
                } else {
                    selected = []
                    if (action.data in state.viewIndex && state.view.length > 1) {
                        let currentIndex = state.viewIndex[action.data],
                            nextSelection
                        if (currentIndex === state.view.length - 1) { // last row
                            nextSelection = state.view[currentIndex - 1]
                        } else {
                            nextSelection = state.view[currentIndex + 1]
                        }
                        selected.push(nextSelection.id)
                    }
                }
            }

            return {
                ...state,
                selected,
                ...store.reduce(state, storeAction)
            }

        case SET_FILTER:
            return {
                ...state,
                filter: action.filter,
                ...store.reduce(state, store.setFilter(makeFilter(action.filter), makeSort(state.sort)))
            }

        case SET_HIGHLIGHT:
            return {
                ...state,
                highlight: action.highlight
            }

        case SET_SORT:
            return {
                ...state,
                sort: action.sort,
                ...store.reduce(state, store.setSort(makeSort(action.sort)))
            }

        case SELECT:
            return {
                ...state,
                selected: action.flowIds
            }

        default:
            return state
    }
}

export function makeFilter(filter?: string): FlowFilterFn | undefined {
    if (!filter) {
        return
    }
    return Filt.parse(filter)
}

export function makeSort({column, desc}: { column?: keyof typeof FlowColumns, desc: boolean }): FlowSortFn {
    if (!column) {
        return (a,b) => 0;
    }
    const sortKeyFun = FlowColumns[column].sortKey
    return (a, b) => {
        const ka = sortKeyFun(a)
        const kb = sortKeyFun(b)
        if (ka > kb) {
            return desc ? -1 : 1
        }
        if (ka < kb) {
            return desc ? 1 : -1
        }
        return 0
    }
}

export function setFilter(filter: string) {
    return {type: SET_FILTER, filter}
}

export function setHighlight(highlight: string) {
    return {type: SET_HIGHLIGHT, highlight}
}

export function setSort(column: string, desc: boolean) {
    return {type: SET_SORT, sort: {column, desc}}
}

export function selectRelative(flows, shift) {
    let currentSelectionIndex = flows.viewIndex[flows.selected[0]]
    let minIndex = 0
    let maxIndex = flows.view.length - 1
    let newIndex
    if (currentSelectionIndex === undefined) {
        newIndex = (shift < 0) ? minIndex : maxIndex
    } else {
        newIndex = currentSelectionIndex + shift
        newIndex = window.Math.max(newIndex, minIndex)
        newIndex = window.Math.min(newIndex, maxIndex)
    }
    let flow = flows.view[newIndex]
    return select(flow ? flow.id : undefined)
}


export function resume(flow: Flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/resume`, {method: 'POST'})
}

export function resumeAll() {
    return dispatch => fetchApi('/flows/resume', {method: 'POST'})
}

export function kill(flow: Flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/kill`, {method: 'POST'})
}

export function killAll() {
    return dispatch => fetchApi('/flows/kill', {method: 'POST'})
}


export function remove(flow: Flow) {
    return dispatch => fetchApi(`/flows/${flow.id}`, {method: 'DELETE'})
}

export function duplicate(flow: Flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/duplicate`, {method: 'POST'})
}

export function replay(flow: Flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/replay`, {method: 'POST'})
}

export function revert(flow: Flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/revert`, {method: 'POST'})
}

export function update(flow: Flow, data) {
    return dispatch => fetchApi.put(`/flows/${flow.id}`, data)
}

export function uploadContent(flow: Flow, file, type) {
    const body = new FormData()
    file = new window.Blob([file], {type: 'plain/text'})
    body.append('file', file)
    return dispatch => fetchApi(`/flows/${flow.id}/${type}/content.data`, {method: 'POST', body})
}


export function clear() {
    return dispatch => fetchApi('/clear', {method: 'POST'})
}

export function upload(file) {
    const body = new FormData()
    body.append('file', file)
    return dispatch => fetchApi('/flows/dump', {method: 'POST', body})
}


export function select(id?: string) {
    return {
        type: SELECT,
        flowIds: id ? [id] : []
    }
}
