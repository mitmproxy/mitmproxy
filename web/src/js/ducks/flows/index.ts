import { assertNever } from "../../utils";

import { Flow } from "../../flow";
import { sortFunctions } from "../../flow/utils";
import { AppDispatch, RootState } from "../store";
import { FilterName } from "../ui/filter";
import { Comparer, createAction, UnknownAction } from "@reduxjs/toolkit";
import {
    buildIndex,
    buildLookup,
    insertViewItem,
    removeViewItemAt,
    updateViewItem,
    withElemRemoved,
} from "./_utils";
import { toSorted, toSpliced } from "./_compat";

export * from "./_backend_actions";

export const FLOWS_ADD = createAction<{
    flow: Flow;
    matching_filters: Partial<{ [key in FilterName]: boolean }>;
}>("flows/add");
export const FLOWS_UPDATE = createAction<{
    flow: Flow;
    matching_filters: Partial<{ [key in FilterName]: boolean }>;
}>("flows/update");
export const FLOWS_REMOVE = createAction<string>("flows/remove");
export const FLOWS_RECEIVE = createAction<Flow[]>("flows/receive");
export const FLOWS_FILTER_UPDATE = createAction<{
    name: FilterName;
    matching_flow_ids: string[] | null;
}>("flows/filterUpdate");

export const setSort = createAction<{
    column?: keyof typeof sortFunctions;
    desc: boolean;
}>("flows/sort");
export const select = createAction<Flow[]>("flows/select");

export interface FlowsState {
    list: Flow[];
    _listIndex: Map<string, number>;
    byId: Map<string, Flow>;
    view: Flow[];
    _viewIndex: Map<string, number>;

    sort: { column?: keyof typeof sortFunctions; desc: boolean };
    selected: Flow[];
    selectedIds: Set<string>;
    highlightedIds: Set<string>;
}

export const defaultState: FlowsState = {
    list: [],
    _listIndex: new Map(),
    byId: new Map(),
    view: [],
    _viewIndex: new Map(),

    sort: { column: undefined, desc: false },
    selected: [],
    selectedIds: new Set(),
    highlightedIds: new Set(),
};

// This is a manual reducer as RTK's createSlice always uses Immer, which is orders of magnitude slower.
export default function flowsReducer(
    state = defaultState,
    action: UnknownAction,
): FlowsState {
    if (FLOWS_RECEIVE.match(action)) {
        const { sort } = state;
        const list = action.payload;
        const _listIndex = buildIndex(list);
        const byId = new Map(list.map((f) => [f.id, f]));
        // No filter information yet, we expect that to come immediately after.
        const view = toSorted(list, makeSort(sort));
        const _viewIndex = buildIndex(view);
        const selected = state.selected
            .map((flow) => byId.get(flow.id))
            .filter((f) => f !== undefined);
        const selectedIds = buildLookup(selected);
        const highlightedIds = new Set<string>();

        return {
            list,
            _listIndex,
            byId,
            view,
            _viewIndex,
            sort,
            selected,
            selectedIds,
            highlightedIds,
        };
    } else if (FLOWS_ADD.match(action)) {
        const { flow, matching_filters } = action.payload;
        const { sort, selected, selectedIds } = state;
        let { view, _viewIndex, highlightedIds } = state;
        // Update list
        const _listIndex = new Map(state._listIndex);
        _listIndex.set(flow.id, state.list.length);
        const list = [...state.list, flow];
        const byId = new Map(state.byId);
        byId.set(flow.id, flow);
        // Update view if filter matches (true) or is unset (undefined).
        if (
            matching_filters[FilterName.Search] === true ||
            matching_filters[FilterName.Search] === undefined
        ) {
            ({ view, _viewIndex } = insertViewItem(
                view,
                _viewIndex,
                flow,
                makeSort(sort),
            ));
        }
        // Update highlight
        if (matching_filters[FilterName.Highlight] === true) {
            highlightedIds = new Set(highlightedIds);
            highlightedIds.add(flow.id);
        }

        return {
            list,
            _listIndex,
            byId,
            view,
            _viewIndex,
            sort,
            selected,
            selectedIds,
            highlightedIds,
        };
    } else if (FLOWS_UPDATE.match(action)) {
        const { flow, matching_filters } = action.payload;
        const { _listIndex, sort, selectedIds } = state;
        let { view, _viewIndex, selected, highlightedIds } = state;
        // Update list
        const listPos = state._listIndex.get(flow.id)!;
        const list = [...state.list];
        list[listPos] = flow;
        const byId = new Map(state.byId);
        byId.set(flow.id, flow);
        // Update view
        const prevViewPos: number | undefined = _viewIndex.get(flow.id);
        const hasOldFlow = prevViewPos !== undefined;
        const hasNewFlow =
            matching_filters[FilterName.Search] === true ||
            matching_filters[FilterName.Search] === undefined;
        if (hasNewFlow && !hasOldFlow) {
            ({ view, _viewIndex } = insertViewItem(
                view,
                _viewIndex,
                flow,
                makeSort(sort),
            ));
        } else if (!hasNewFlow && hasOldFlow) {
            ({ view, _viewIndex } = removeViewItemAt(
                view,
                _viewIndex,
                prevViewPos,
            ));
        } else if (hasNewFlow && hasOldFlow) {
            ({ view, _viewIndex } = updateViewItem(
                view,
                _viewIndex,
                flow,
                makeSort(sort),
            ));
        }
        // Update selection
        if (selectedIds.has(flow.id)) {
            selected = selected.map((existing) =>
                existing.id === flow.id ? flow : existing,
            );
        }
        // Update highlight
        if (matching_filters[FilterName.Highlight] === true) {
            if (!highlightedIds.has(flow.id)) {
                highlightedIds = new Set(highlightedIds);
                highlightedIds.add(flow.id);
            }
        } else {
            highlightedIds = withElemRemoved(highlightedIds, flow.id);
        }

        return {
            list,
            _listIndex,
            byId,
            view,
            _viewIndex,
            sort,
            selected,
            selectedIds,
            highlightedIds,
        };
    } else if (FLOWS_REMOVE.match(action)) {
        const flow_id = action.payload;
        const { sort } = state;
        let { view, _viewIndex, selected, selectedIds } = state;
        const listPos = state._listIndex.get(flow_id)!;
        const list = toSpliced(state.list, listPos, 1);
        const _listIndex = buildIndex(list);
        const byId = new Map(state.byId);
        byId.delete(flow_id);
        // Update view
        const viewPos = _viewIndex.get(flow_id);
        if (viewPos !== undefined) {
            ({ view, _viewIndex } = removeViewItemAt(
                view,
                _viewIndex,
                viewPos,
            ));
        }
        // Update selection
        if (selectedIds.has(flow_id)) {
            if (selected.length === 1 && viewPos !== undefined) {
                const fallback =
                    view[viewPos /* no +1, already removed */] ??
                    view[viewPos - 1];
                selected = fallback ? [fallback] : [];
            } else {
                selected = selected.filter((f) => f.id !== flow_id);
            }
            selectedIds = buildLookup(selected);
        }
        // Update highlight
        const highlightedIds = withElemRemoved(state.highlightedIds, flow_id);

        return {
            list,
            _listIndex,
            byId,
            view,
            _viewIndex,
            sort,
            selected,
            selectedIds,
            highlightedIds,
        };
    } else if (FLOWS_FILTER_UPDATE.match(action)) {
        const { name, matching_flow_ids } = action.payload;
        switch (name) {
            case FilterName.Search: {
                const view = toSorted(
                    matching_flow_ids === null
                        ? state.list
                        : matching_flow_ids.map((id) => state.byId.get(id)!),
                    makeSort(state.sort),
                );
                const _viewIndex = buildIndex(view);
                return {
                    ...state,
                    view,
                    _viewIndex,
                };
            }
            case FilterName.Highlight:
                return {
                    ...state,
                    highlightedIds: new Set(matching_flow_ids),
                };
            /* istanbul ignore next @preserve */
            default:
                assertNever(name);
        }
    } else if (setSort.match(action)) {
        const sort = action.payload;
        let view: Flow[];
        if (sort.column) {
            view = toSorted(state.view, makeSort(sort));
        } else {
            // Restore original order if column isn't specified.
            view = state.list.filter((f) => state._viewIndex.has(f.id));
        }
        const _viewIndex = buildIndex(view);
        return {
            ...state,
            sort,
            view,
            _viewIndex,
        };
    } else if (select.match(action)) {
        return {
            ...state,
            selected: action.payload,
            selectedIds: buildLookup(action.payload),
        };
    } else {
        return state;
    }
}

export function makeSort({
    column,
    desc,
}: {
    column?: keyof typeof sortFunctions;
    desc: boolean;
}): Comparer<Flow> {
    if (!column) {
        return (_a, _b) => 0;
    }
    const sortKeyFun = sortFunctions[column];
    return (a, b) => {
        const ka = sortKeyFun(a);
        const kb = sortKeyFun(b);
        // @ts-expect-error undefined is fine
        if (ka > kb) {
            return desc ? -1 : 1;
        }
        // @ts-expect-error undefined is fine
        if (ka < kb) {
            return desc ? 1 : -1;
        }
        return 0;
    };
}

export function selectRelative(flows: FlowsState, shift: number) {
    const lastSelected: Flow | undefined =
        flows.selected[flows.selected.length - 1];
    const currentSelectionIndex: number | undefined = flows._viewIndex.get(
        lastSelected?.id,
    );
    const minIndex = 0;
    const maxIndex = flows.view.length - 1;
    let newIndex: number;
    if (currentSelectionIndex === undefined) {
        newIndex = shift < 0 ? minIndex : maxIndex;
    } else {
        newIndex = currentSelectionIndex + shift;
        newIndex = window.Math.max(newIndex, minIndex);
        newIndex = window.Math.min(newIndex, maxIndex);
    }
    const flow = flows.view[newIndex];
    return select(flow ? [flow] : []);
}

/** Toggle selection for one particular flow. */
export function selectToggle(flow: Flow) {
    return (dispatch: AppDispatch, getState: () => RootState) => {
        const { flows } = getState();
        if (flows.selectedIds.has(flow.id)) {
            dispatch(select(flows.selected.filter((f) => f !== flow)));
        } else {
            dispatch(select([...flows.selected, flow]));
        }
    };
}

/** Select a range of flows with shift + click. */
export function selectRange(flow: Flow) {
    return (dispatch: AppDispatch, getState: () => RootState) => {
        const { flows } = getState();
        const prev = flows.selected[flows.selected.length - 1];

        const thisIndex = flows._viewIndex.get(flow.id);
        const prevIndex = flows._viewIndex.get(prev?.id);
        if (thisIndex === undefined || prevIndex === undefined) {
            return dispatch(select([flow]));
        }
        let newSelection: Flow[];
        if (thisIndex <= prevIndex) {
            newSelection = flows.view.slice(thisIndex, prevIndex + 1);
        } else {
            newSelection = flows.view.slice(prevIndex + 1, thisIndex + 1);
            newSelection.push(prev); // Keep this at the end if the user shift-clicks again.
        }
        return dispatch(select(newSelection));
    };
}
