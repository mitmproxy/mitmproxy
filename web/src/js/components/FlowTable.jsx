import React from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import { connect } from 'react-redux'
import shallowEqual from 'shallowequal'
import AutoScroll from './helpers/AutoScroll'
import { calcVScroll } from './helpers/VirtualScroll'
import FlowTableHead from './FlowTable/FlowTableHead'
import FlowRow from './FlowTable/FlowRow'
import Filt from "../filt/filt"
import * as flowsActions from '../ducks/flows'


class FlowTable extends React.Component {

    static propTypes = {
        selectFlow: PropTypes.func.isRequired,
        flows: PropTypes.array.isRequired,
        rowHeight: PropTypes.number,
        highlight: PropTypes.string,
        selected: PropTypes.object,
    }

    static defaultProps = {
        rowHeight: 32,
    }

    constructor(props, context) {
        super(props, context)

        this.state = { vScroll: calcVScroll() }
        this.onViewportUpdate = this.onViewportUpdate.bind(this)
    }

    UNSAFE_componentWillMount() {
        window.addEventListener('resize', this.onViewportUpdate)
    }

    UNSAFE_componentWillUnmount() {
        window.removeEventListener('resize', this.onViewportUpdate)
    }

    componentDidUpdate() {
        this.onViewportUpdate()

        if (!this.shouldScrollIntoView) {
            return
        }

        this.shouldScrollIntoView = false

        const { rowHeight, flows, selected } = this.props
        const viewport = ReactDOM.findDOMNode(this)
        const head = ReactDOM.findDOMNode(this.refs.head)

        const headHeight = head ? head.offsetHeight : 0

        const rowTop = (flows.indexOf(selected) * rowHeight) + headHeight
        const rowBottom = rowTop + rowHeight

        const viewportTop = viewport.scrollTop
        const viewportHeight = viewport.offsetHeight

        // Account for pinned thead
        if (rowTop - headHeight < viewportTop) {
            viewport.scrollTop = rowTop - headHeight
        } else if (rowBottom > viewportTop + viewportHeight) {
            viewport.scrollTop = rowBottom - viewportHeight
        }
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        if (nextProps.selected && nextProps.selected !== this.props.selected) {
            this.shouldScrollIntoView = true
        }
    }

    onViewportUpdate() {
        const viewport = ReactDOM.findDOMNode(this)
        const viewportTop = viewport.scrollTop || 0

        const vScroll = calcVScroll({
            viewportTop,
            viewportHeight: viewport.offsetHeight || 0,
            itemCount: this.props.flows.length,
            rowHeight: this.props.rowHeight,
        })

        if (this.state.viewportTop !== viewportTop || !shallowEqual(this.state.vScroll, vScroll)) {
            this.setState({ vScroll, viewportTop })
        }
    }

    render() {
        const { vScroll, viewportTop } = this.state
        const { flows, selected, highlight } = this.props
        const isHighlighted = highlight ? Filt.parse(highlight) : () => false

        return (
            <div className="flow-table" onScroll={this.onViewportUpdate}>
                <table>
                    <thead ref="head" style={{ transform: `translateY(${viewportTop}px)` }}>
                        <FlowTableHead />
                    </thead>
                    <tbody>
                        <tr style={{ height: vScroll.paddingTop }}/>
                        {flows.slice(vScroll.start, vScroll.end).map(flow => (
                            <FlowRow
                                key={flow.id}
                                flow={flow}
                                selected={flow === selected}
                                highlighted={isHighlighted(flow)}
                                onSelect={this.props.selectFlow}
                            />
                        ))}
                        <tr style={{ height: vScroll.paddingBottom }}/>
                    </tbody>
                </table>
            </div>
        )
    }
}

export const PureFlowTable = AutoScroll(FlowTable)

export default connect(
    state => ({
        flows: state.flows.view,
        highlight: state.flows.highlight,
        selected: state.flows.byId[state.flows.selected[0]],
    }),
    {
        selectFlow: flowsActions.select,
    }
)(PureFlowTable)
