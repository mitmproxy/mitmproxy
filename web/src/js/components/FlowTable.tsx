import * as React from "react";
import { connect, shallowEqual } from "react-redux";
import * as autoscroll from "./helpers/AutoScroll";
import { calcVScroll, VScroll } from "./helpers/VirtualScroll";
import FlowTableHead from "./FlowTable/FlowTableHead";
import FlowRow from "./FlowTable/FlowRow";
import Filt from "../filt/filt";
import { Flow } from "../flow";
import { RootState } from "../ducks";

type FlowTableProps = {
    flows: Flow[];
    rowHeight: number;
    highlight: string;
    selected: Flow;
};

type FlowTableState = {
    vScroll: VScroll;
    viewportTop: number;
};

export class PureFlowTable extends React.Component<
    FlowTableProps,
    FlowTableState
> {
    static defaultProps = {
        rowHeight: 32,
    };
    private viewport = React.createRef<HTMLDivElement>();
    private head = React.createRef<HTMLTableSectionElement>();

    constructor(props, context) {
        super(props, context);

        this.state = {
            vScroll: calcVScroll(),
            viewportTop: 0,
        };
        this.onViewportUpdate = this.onViewportUpdate.bind(this);
    }

    componentDidMount() {
        window.addEventListener("resize", this.onViewportUpdate);
        this.onViewportUpdate();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.onViewportUpdate);
    }

    getSnapshotBeforeUpdate() {
        return autoscroll.isAtBottom(this.viewport);
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (snapshot) {
            autoscroll.adjustScrollTop(this.viewport);
        }
        this.onViewportUpdate();

        const selectedNewFlow =
            this.props.selected && this.props.selected !== prevProps.selected;
        if (selectedNewFlow) {
            const { rowHeight, flows, selected } = this.props;
            const viewport = this.viewport.current!;
            const head = this.head.current;

            const headHeight = head ? head.offsetHeight : 0;

            const rowTop = flows.indexOf(selected) * rowHeight + headHeight;
            const rowBottom = rowTop + rowHeight;

            const viewportTop = viewport.scrollTop;
            const viewportHeight = viewport.offsetHeight;

            // Account for pinned thead
            if (rowTop - headHeight < viewportTop) {
                viewport.scrollTop = rowTop - headHeight;
            } else if (rowBottom > viewportTop + viewportHeight) {
                viewport.scrollTop = rowBottom - viewportHeight;
            }
            this.onViewportUpdate();
        }
    }

    onViewportUpdate() {
        const viewport = this.viewport.current!;
        const viewportTop = viewport.scrollTop || 0;

        const vScroll = calcVScroll({
            viewportTop,
            viewportHeight: viewport.offsetHeight || 0,
            itemCount: this.props.flows.length,
            rowHeight: this.props.rowHeight,
        });

        if (
            this.state.viewportTop !== viewportTop ||
            !shallowEqual(this.state.vScroll, vScroll)
        ) {
            // the next rendered state may only have much lower number of rows compared to what the current
            // viewportHeight anticipates. To make sure that we update (almost) at once, we already constrain
            // the maximum viewportTop value. See https://github.com/mitmproxy/mitmproxy/pull/5658 for details.
            const newViewportTop = Math.min(
                viewportTop,
                vScroll.end * this.props.rowHeight,
            );
            this.setState({
                vScroll,
                viewportTop: newViewportTop,
            });
        }
    }

    render() {
        const { vScroll, viewportTop } = this.state;
        const { flows, selected, highlight } = this.props;
        const isHighlighted = highlight ? Filt.parse(highlight) : () => false;

        return (
            <div
                className="flow-table"
                onScroll={this.onViewportUpdate}
                ref={this.viewport}
            >
                <table>
                    <thead
                        ref={this.head}
                        style={{ transform: `translateY(${viewportTop}px)` }}
                    >
                        <FlowTableHead />
                    </thead>
                    <tbody>
                        <tr style={{ height: vScroll.paddingTop }} />
                        {flows.slice(vScroll.start, vScroll.end).map((flow) => (
                            <FlowRow
                                key={flow.id}
                                flow={flow}
                                selected={flow === selected}
                                highlighted={isHighlighted(flow)}
                            />
                        ))}
                        <tr style={{ height: vScroll.paddingBottom }} />
                    </tbody>
                </table>
            </div>
        );
    }
}

export default connect((state: RootState) => ({
    flows: state.flows.view,
    highlight: state.flows.highlight,
    selected: state.flows.byId[state.flows.selected[0]],
}))(PureFlowTable);
