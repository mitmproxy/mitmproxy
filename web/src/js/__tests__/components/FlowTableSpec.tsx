import * as React from "react";
import FlowTable, { PureFlowTable } from "../../components/FlowTable";

import { act, render } from "../test-utils";
import { select } from "../../ducks/flows";

window.addEventListener = jest.fn();

describe("FlowTable Component", () => {
    it("should render correctly", () => {
        const { asFragment } = render(<FlowTable />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should scroll current selection into view", () => {
        const height = PureFlowTable.defaultProps.rowHeight;
        const { asFragment, store } = render(
            <div style={{ height, overflow: "hidden" }}>
                <FlowTable />
            </div>,
        );
        expect(asFragment()).toMatchSnapshot();

        act(() => store.dispatch(select(store.getState().flows.view[3].id)));
        expect(asFragment()).toMatchSnapshot();
    });
});
