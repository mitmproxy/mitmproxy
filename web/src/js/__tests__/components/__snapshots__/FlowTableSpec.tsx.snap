// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowTable Component should render correctly 1`] = `
<div
  className="flow-table"
  onScroll={[Function]}
>
  <table>
    <thead
      style={
        Object {
          "transform": "translateY(0px)",
        }
      }
    >
      <tr>
        <th
          className="col-tls"
          onClick={[Function]}
        >
          
        </th>
        <th
          className="col-icon"
          onClick={[Function]}
        >
          
        </th>
        <th
          className="col-path sort-desc"
          onClick={[Function]}
        >
          Path
        </th>
        <th
          className="col-method"
          onClick={[Function]}
        >
          Method
        </th>
        <th
          className="col-status"
          onClick={[Function]}
        >
          Status
        </th>
        <th
          className="col-size"
          onClick={[Function]}
        >
          Size
        </th>
        <th
          className="col-time"
          onClick={[Function]}
        >
          Time
        </th>
        <th
          className="col-quickactions"
          onClick={[Function]}
        >
          
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        style={
          Object {
            "height": 0,
          }
        }
      />
      <tr
        className="has-request has-response"
        onClick={[Function]}
      >
        <td
          className="col-tls col-tls-https"
        />
        <td
          className="col-icon"
        >
          <div
            className="resource-icon resource-icon-websocket"
          />
        </td>
        <td
          className="col-path"
        >
          <i
            className="fa fa-fw fa-exclamation pull-right"
          />
          <span
            className="marker pull-right"
          >
            
          </span>
          http://address:22/path
        </td>
        <td
          className="col-method"
        >
          WSS
        </td>
        <td
          className="col-status"
          style={
            Object {
              "color": "darkgreen",
            }
          }
        >
          200
        </td>
        <td
          className="col-size"
        >
          43b
        </td>
        <td
          className="col-time"
        >
          5s
        </td>
        <td
          className="col-quickactions"
          onClick={[Function]}
        />
      </tr>
      <tr
        style={
          Object {
            "height": 0,
          }
        }
      />
    </tbody>
  </table>
</div>
`;
